### ISSUE TITLE
Regression: Spurious failure with Caffeine 3.1.2

### ISSUE THREAD

**Issue by henrik242** [ben-manes/caffeine#831]
I have a Kotlin object with: 
```
import com.github.benmanes.caffeine.cache.*
import com.github.benmanes.caffeine.cache.Caffeine.newBuilder
import com.github.benmanes.caffeine.cache.Ticker.systemTicker
import java.time.Duration

object MyCache {
    @JvmStatic
    @JvmOverloads
    fun <T, R> createCache(maxSize: Int, expiration: Duration, ticker: Ticker = systemTicker()): Cache<T, R> =
        newBuilder()
            .ticker(ticker)
            .maximumSize(maxSize.toLong())
            .expireAfterWrite(expiration)
            .build()
}
```
 ... and a Groovy/Spock test:
```
class MyTest extends Specification {

    def "a full cache should throw away old stuff and make room for new results when get() is called"() {
      given: "a very full cache"
        def cache = MyCache.createCache(3, ofDays(7))

        for (int i = 0; i < 10; i++) {
            cache.put("key${i}", "value${i}")
        }

      when: "we fetch a non-existant value, programmatically setting a default value"
        def foo = cache.get("created-by-get") { "honey, i'm home" }
        def contents = cache.asMap()

      then:
        foo == "honey, i'm home"
        contents["created-by-get"] == "honey, i'm home"
        contents.size() == 3
    }
}
```
This always succeeded in Caffeine 3.1.1, now it fails randomly.

**Comment by ben-manes** [comment]
If you set `Caffeine.executor(Runnable::run)` then the eviction will be performed on the calling thread and your test passes. The admission window is sized at 1 and main region at 2, with key9 pushed out of the window as a candidate. That leaves "created-by-get" as protected within the admission window, so it favors keeping the most recent entry.

In your test, however, it is configured by default to evict asynchronously. That means the test thread races and might observe a cache size much greater than 3 and need to evict multiple entries. When that occurs it has to walk the regions, which causes it to more aggressively evaluate the admission window. I'd need to dig in more to figure out what caused it to decide on the MRU and if there is a mistake, but none of this is intended to be strongly predictable. There are probabilistic decisions such as frequency estimates, hash flooding protection, etc which makes it less predictable. For now I think changing your test to not use an async executor would be better, e.g. even prior the test runner could have raced to see a size greater than 3 without a wait condition.

**Comment by ben-manes** [comment]
When the window overflows then it evicts in LRU order its items into the main region's probation list, placed into its MRU position. As multiple items are evicted, a sublist is promoted and the first is captured for a candidate traversal. If the main region needs to evict, it compares the candidates (MRU sublist) to the LRU victims, using a frequency estimate to decide which is more valuable to retain (TinyLFU). In a tiny cache these two traversals can overlap and point to the same element, which would leave one pointer invalidated when that entry is evicted. When this case is detected, the candidate pointer switches to the admission window to traverse it from LRU order. As the cache might have varying weights or the maximum size changed dynamically, it has to use a more complex search to be able to fully walk the contents in case it greatly exceeds the capacity (e.g. a heavy item inserted, maxSize set to zero).

In your test is that due to the eviction running async, the test inserts all 10 items before the eviction occurs. These inserts go into a write buffer, sized at 128 * num_cpus, so there is enough slack to accept all at once (or else it blocks writes to ensure eviction can keep up). When eviction runs it then promotes 9 entries into the main region, but since all came from the window the main's victim and first promoted victim are the same. This causes it to flip the search to the window, which makes your newest entry eligible for eviction. It then compares "key0" vs "created-by-get", and since they have the same frequency score the window item is evicted. This is because we bias those entries that are in the main region as usually they are promoted due to having previously won the frequency comparison, so other than a warm-up it has a higher likelihood to be reused.

In this warm-up case, we run into a scenario where any of the items is reasonable victim as all are one-hit wonders (poor frequency). We might bias towards recency, but at such a tiny capacity that is logically equivalent. Therefore any item is a reasonable victim as they have the same likelihood of reuse, as there are no good signals to infer from. If we prefer recency on a tie breaker then we would lose scan resistance (e.g. [LRU loop](https://github.com/ben-manes/caffeine/wiki/Efficiency#glimpse)). Instead we favor whatever was promoted by hoping that happened for a good reason after a warm-up, but it does allow a warm up to get a small artificial bias. Once there is activity or the cache is at least modestly sized then these observations go away.

As the policy is a mixture of recency and frequency heuristics, I think this is an okay compromise. From its perspective all items are equally valuable and too strong of a recency preference is a negative. After usage the frequency can be leveraged, the window/main regions will dynamically resize, etc. to maximize the long-term hit rate. At the very start there is not enough history to make good decisions, but it's also not worth complicating to make human predictable as such a short part of the application's runtime. It is mostly observable in tests where we have our human assumptions, but those are not mathematically any better.

In summary, I think this is okay and I don't want to optimize for (already broken) unit tests. If you observe a degradation in performance for a long running process that is worth fixing, but for this particular case I think it is okay. Any tweaks to the policy can cause little changes like you observed, but if the long running hit rate stays high then it is acceptable noise in the policy's predictions.

**Comment by henrik242** [comment]
Thanks for that deep dive into the caffeine innards! Very interesting and much appreciated. We'll change our unit test accordingly :)
